### PR TITLE
fix(observability): group per-repo Sentry noise; stop reporting 502s as authz denials

### PR DESCRIPTION
**Pull request by jon-bell**
Triage of the Bugsink **Dev** project (where the e2e suites report): 744 issues / 497k events since 2025-09-10, of which 114 have been seen since July 1. Two problems, both of which made the project harder to read than the bugs in it.

## Noise: one defect arriving as N issues

Many messages embed the thing they failed on — a repo name, a sha, a queue read count — and Bugsink hashes the message into the group key. A single run of the push-webhook suite produced **72 issues that were all one line of code**:

> `Could not resolve the current head of pawtograder-playground/test-e2e-student-repo--msh98vk8kvtq7w to check whether deadbeefmsh98vk8kvtq7w is superseded (…)`

and `No octokit found for <repo>` had accumulated 29.

New `_shared/SentryFingerprint.ts` is a `beforeSend` hook that normalizes volatile identifiers (repo slugs, shas, uuids, per-run ids, numbers) out of the message and sets an explicit fingerprint of `[type, innermost in-app frame, normalized message]`. Wired into all 13 `Sentry.init` sites.

Measured over all 744 real issues pulled from the API: **268 distinct groups → 151**. The two cases above collapse to 2 (the `Error` and `AggregateError` wrappers stay separate — different types) and 1.

Messages themselves are untouched, so each event still names its own repo and sha; only the grouping key changes. Two deliberate guards:

- An explicit call-site fingerprint always wins — the rate-limit sites in `github-async-worker` and friends group by installation-independent keys on purpose.
- A message with nothing to normalize keeps Sentry's default grouping, rather than detaching from its existing issue history.
- Octokit doc links (`… - https://docs.github.com/rest/repos/repos#get-a-repository`) survive intact, since they identify which call failed.

## Errors: 502s reported to users as permission denials

The authorization assertions could not tell *"the answer is no"* from *"the question could not be asked"*. `assertUserIsInCourse` discarded the query error entirely, so a 502 from Kong on the `user_roles` lookup told an **enrolled** student `SecurityError` → 401, "not enrolled in this course". `mcp-tokens` answered "MCP tokens are only available to instructors and graders" on a 502, and "Unauthorized" when `auth/v1/user` 502'd.

Nothing retries a 401, and the resulting Sentry events are indistinguishable from the genuine denials the negative-path e2e tests produce — so the failure hid inside expected noise. That is how it went unnoticed.

A failed lookup is now a retryable **503**. An empty result keeps its existing status, and `PGRST116` from `.single()` is treated as the denial it is, so **no negative-path behavior changes**. Fixed in `assertUserIsInstructor`, `assertUserIsAdmin`, `assertUserIsInstructorOrGrader`, `assertUserIsInCourse`, and both `mcp-tokens` checks.

## Verification

- 18 new unit tests (10 fingerprint, 8 authorization), pinned to verbatim messages from the project. All 67 `_shared` Deno tests pass.
- `deno check` on the touched files reports the same **13** pre-existing octokit webhook-type errors as staging — zero delta.
- Prettier clean. Nothing outside `supabase/functions/` is touched.

## Not included, and why

- **The 502s themselves** are a Kong/PostgREST capacity problem under e2e load, not a code fault. This PR stops them from lying to users; it does not fix their cause. They also surface in `gradebook-column-recalculate` (34k events), submission-file ingestion, and `enrollments-add`, which is worth its own look.
- **The two largest real bugs in the triage were already fixed on staging** — `sync_staff_team` `.single()` on zero rows (#880) and the `workflow_run_error` FK violation (#884). Their remaining events come from e2e runs on branches cut before those merges, and should stop once branches rebase. Worth confirming rather than re-investigating.
- **Orphaned submission when the compensating cleanup also 502s** is being handled on `feat/895-repo-only-no-autograder`, so it is left to that branch.
- The 378-issue `Failed to fetch JWKS` group (stale, last seen 2025-10-17) does **not** collapse — those split on *stacktrace*, not message, so message normalization cannot reach them.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: supabase/functions/_shared/HandlerUtils.ts [hunk 1]
```diff
@@ -158,6 +165,32 @@ export async function assertUserIsInstructorOrServiceRole(courseId: number, auth
...
+ */
+export function assertRoleLookupSucceeded(error: { code?: string; message: string } | null, operation: string): void {
+  if (!error || error.code === "PGRST116") return;
+  throw new UserVisibleError(`${operation} is temporarily unavailable: ${error.message}`, 503);
+}
+
+/** Same, for the auth server. Only an explicit 4xx is a statement about the token. */
+export function assertAuthLookupSucceeded(error: { status?: number; message: string } | null | undefined): void {
+  if (!error) return;
+  if (error.status !== undefined && error.status < 500) return;
```
[the remark this answers is not in the document]
Comment (jon-bell) [reply supabase/functions/_shared/HandlerUtils.ts]: Fixed in 002f7786 — good catch, this was the one finding that reintroduced the exact bug the PR set out to remove.

`AuthRetryableFetchError` does carry `status: 0`, so `error.status !== undefined && error.status < 500` accepted it as a real answer from the auth server and fell through to `SecurityError` → 401. Inverted the test to check for an explicit 4xx instead, so `0`, absent, 3xx and 5xx all raise 503:

```ts
if (error.status !== undefined && error.status >= 400 && error.status < 500) return;
throw new UserVisibleError(`Authentication is temporarily unavailable: ${error.message}`, 503);
```

Applied the same inversion to the equivalent predicate in `mcp-tokens/index.ts:98`. Covered by two new tests in `_shared/HandlerUtils.test.ts` — one pinning `status: 0` specifically, one sweeping 301/500/502/504.

File: supabase/functions/_shared/SentryFingerprint.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  const frame = [...frames].reverse().find((f) => f.in_app) ?? frames[frames.length - 1];
+  const file = frame.filename?.split("/").slice(-2).join("/") ?? "?";
+  return `${file}:${frame.function ?? "?"}`;
+}
+
+function eventMessage(event: FingerprintableEvent): string {
+  const exc = event.exception?.values?.[0];
+  if (exc?.value) return exc.value;
+  if (typeof event.message === "string") return event.message;
+  if (event.message) return event.message.formatted ?? event.message.message ?? "";
```
[the remark this answers is not in the document]
Comment (jon-bell) [reply supabase/functions/_shared/SentryFingerprint.ts:107]: Fixed in 002f7786. Confirmed against a real event before changing it: issue `fbbf860b` in the Dev project has `values = [Error, AggregateError]` for a captured `AggregateError`, and Bugsink’s own `calculated_type` for that issue is `AggregateError` — so the captured exception is indeed last.

Added a `capturedException()` helper that reads `values[values.length - 1]`, and routed all three uses through it (type, `topFrameKey`, and `eventMessage`). Side benefit you implied: our fingerprint’s type now agrees with the type Bugsink displays.

New test `linked errors fingerprint on the captured wrapper, not its first child` varies both the child’s type and its innermost frame and asserts the fingerprint is unchanged. Also added a companion test for the case where the wrapper message has nothing to normalize — that correctly returns `null` and keeps default grouping, since Bugsink already keys on the captured exception there.

File: supabase/functions/_shared/SentryFingerprint.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    message
+      // uuids: profile/class/message ids interpolated into messages
+      .replace(/\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/gi, "<uuid>")
+      // owner/repo slugs, including the e2e suffixes (test-e2e-student-repo--msh98vk8kvtq7w).
+      // URLs lead the alternation so the octokit doc links appended to every HttpError message
+      // ("… - https://docs.github.com/rest/repos/repos#get-a-repository") survive intact instead of
+      // being shredded into <repo> fragments — they are constant, and they name the failing call.
+      .replace(/https?:\/\/\S+|\b[\w.-]+\/[\w.-]+\b/g, (m) => (m.startsWith("http") ? m : "<repo>"))
+      // git shas: real (7-40 hex). A digit is required so ordinary a-f words ("defaced") are not
+      // mistaken for one. The e2e fakes (deadbeefmsh98vk8kvtq7w) are not hex and fall to <id> below.
```
[the remark this answers is not in the document]
Comment (jon-bell) [reply supabase/functions/_shared/SentryFingerprint.ts:58]: Fixed in 002f7786, with one adjustment worth flagging.

Status codes are now preserved as `<status:NNN>` before the catch-all, so your `GitHubSyncHelpers.ts:766` example keeps 404 and 500 distinct. React error codes (`#418` vs `#423`) benefit the same way.

The adjustment: I checked the 744-issue corpus first, and a bare 4xx/5xx range match would have made things worse in a way that is not obvious — the same digit ranges occur inside embedded stack traces (`chunks/476.js`, `26_fetch.js:485:11`), so preserving them would have *split* one group across every build. The rule is therefore positional: a code counts only as a free-standing token, not touching `.`, `:`, `/`, or another digit.

```ts
.replace(/(^|[\s#([])([45]\d{2})(?=$|[\s),;\]]|\.(?:\s|$))/g, "$1<status:$2>")
.replace(/<status:\d{3}>|\d+/g, (m) => (m.startsWith("<status:") ? m : "<n>"))
```

That second line matters too: on the first pass the placeholder’s own digits were being eaten by the catch-all, yielding `<status:<n>>` and defeating the fix — so the token has to lead the alternation. Both directions are tested (`HTTP status codes survive normalization…` and `status-shaped digits inside stack traces are still collapsed`), and the corpus still collapses 268 → 151.

File: supabase/functions/mcp-tokens/index.ts [hunk 4]
```diff
@@ -37,6 +39,24 @@ const corsHeaders = {
...
+ * The response status here is derived from the thrown message, so a lookup that merely *failed* used
+ * to land on the same branch as a lookup that came back empty: a 502 from PostgREST on the role query
+ * answered the user "MCP tokens are only available to instructors and graders", and a 502 from
+ * auth/v1/user answered "Unauthorized". Both are wrong and both are unactionable — the caller retries
+ * nothing because it was told the answer is no. Instructors saw this during e2e runs while Kong was
+ * returning 502s. Raising this instead maps to 503, which is true and which clients retry.
+ */
+class UpstreamUnavailableError extends Error {
+  constructor(operation: string, cause: unknown) {
+    const detail = cause instanceof Error ? cause.message : String(cause);
```
[the remark this answers is not in the document]
Comment (jon-bell) [reply supabase/functions/mcp-tokens/index.ts]: Fixed in 002f7786. Confirmed the shape from a real event first — `{code: "PGRST116", details: "The result contains 0 rows", hint: null, message: "..."}` — so this was rendering as `[object Object]` exactly as you describe.

Extracted `describeCause` into `_shared/ErrorDetail.ts` rather than inlining it: the function lives behind `Deno.serve` in `mcp-tokens/index.ts`, so it was untestable there, and the same need will come up elsewhere. It reads `message` off anything carrying one, appends the PostgREST `code` when present, and falls back to `JSON.stringify` (guarded for circular refs) instead of ever emitting `[object Object]`.

Also took the second half of your suggestion — the wrapper now retains `cause`, so the Sentry event keeps the code/details/hint. Seven tests in `_shared/ErrorDetail.test.ts`, including the verbatim PostgREST object.
